go-controller/pkg/ovn/controller/services/services_controller.go

```diff
@@ -0,0 +1,484 @@
+					}
+				}
+			}
+			// Services ExternalIPs and LoadBalancer.IngressIPs have the same behavior in OVN
+			// so they are aggregated in a slice and processed together.
+			var externalIPs []string
+			// ExternalIP
+			for _, extIP := range service.Spec.ExternalIPs {
+				// only use the IPs of the same ClusterIP family
+				if utilnet.IsIPv6String(extIP) == utilnet.IsIPv6String(ip) {
+					externalIPs = append(externalIPs, extIP)
+				}
+			}
+			// LoadBalancer
+			for _, ingress := range service.Status.LoadBalancer.Ingress {
+				// only use the IPs of the same ClusterIP family
+				if ingress.IP != "" && utilnet.IsIPv6String(ingress.IP) == utilnet.IsIPv6String(ip) {
+					externalIPs = append(externalIPs, ingress.IP)
+				}
+			}
+
+			// reconcile external IPs
+			if len(externalIPs) > 0 {
+				gatewayRouters, _, err := gateway.GetOvnGateways()
+				if err != nil {
+					return err
+				}
+				for _, extIP := range externalIPs {
+					// only use the IPs of the same ClusterIP family
+					if utilnet.IsIPv6String(extIP) != utilnet.IsIPv6String(ip) {
+						continue
+					}
+					for _, gatewayRouter := range gatewayRouters {
+						lbID, err := gateway.GetGatewayLoadBalancer(gatewayRouter, svcPort.Protocol)
+						if err != nil {
+							klog.Warningf("Service Sync: Gateway router %s does not have load balancer (%v)",
+								gatewayRouter, err)
+							// TODO: why continue? should we error and requeue and retry?
+							continue
+						}
+						vip := util.JoinHostPortInt32(extIP, svcPort.Port)
+						// Reconcile OVN
+						klog.V(4).Infof("Updating ExternalIP service %s/%s with VIP %s %s", name, namespace, vip, svcPort.Protocol)
+						if err := loadbalancer.UpdateLoadBalancer(lbID, vip, eps); err != nil {
+							c.eventRecorder.Eventf(service, v1.EventTypeWarning, "FailedToUpdateOVNLoadBalancer",
+								"Error trying to update OVN LoadBalancer for Service %s/%s: %v", name, namespace, err)
+							return err
+						}
+						c.serviceTracker.updateService(name, namespace, vip, svcPort.Protocol)
+					}
+				}
+			}
+		}
+	}
+	return nil
+}
+
+// handlers
+
+// onServiceUpdate queues the Service for processing.
+func (c *Controller) onServiceAdd(obj interface{}) {
+	key, err := cache.MetaNamespaceKeyFunc(obj)
+	if err != nil {
+		utilruntime.HandleError(fmt.Errorf("Couldn't get key for object %+v: %v", obj, err))
+		return
+	}
+	klog.V(4).Infof("Adding service %s", key)
+	c.queue.Add(key)
+}
+
+// onServiceUpdate updates the Service Selector in the cache and queues the Service for processing.
+func (c *Controller) onServiceUpdate(oldObj, newObj interface{}) {
+	oldService := oldObj.(*v1.Service)
+	newService := newObj.(*v1.Service)
+
+	// don't process resync or objects that are marked for deletion
+	if oldService.ResourceVersion == newService.ResourceVersion ||
+		!newService.GetDeletionTimestamp().IsZero() {
+		return
+	}
+
+	key, err := cache.MetaNamespaceKeyFunc(newObj)
```

Comment:
Question here: what happens if a service changes IP or port? I didn't see any cleanup of that in `syncServices`, doesn't that mean that we will continue programming OVN with the new VIP (new IP/new port) and leave the stale item in the DB. I guess something slipped me by...

Reply:
OK, I think I get it...it's being done in the next commit with the "repair loop". Haha, this PR is toooo big :D

Reply:
OK, but wait a minute...now I see in the third commit after that, that the repair loop is only run once at statup? So then how do we handle my initial question?

Reply:
The tracker is in charge of tracking the services for deletion
 > what happens if a service changes IP or port
I have to double check, but AFAIK this is not possible:
-  ClusterIP in inmutable once set, but on serviceType change, i.e ClusterIP -> ExternalName) something we are taking into consideration.
- I have doubts about service ports, they are a list,  but almost sure you don´ t modify them, you create or delete
cc: @danwinship

Reply:
service ports are mutable; `syncServices` needs to make sure that it cleans up any removed/changed ports

Reply:
> service ports are mutable; `syncServices` needs to make sure that it cleans up any removed/changed ports

heh, I have a bug here

Reply:
fixed the bug, need to add unit tests. 
My logic to fix this is, because we track the VIPs that a service has, once we process the service object, those vips that remain in the tracker are the ones that mutated or were deleted.

Reply:
added a new test here https://github.com/ovn-org/ovn-kubernetes/pull/1857/commits/253efee272e4fde9d9b2ba4917d61dcb89ea11cd#diff-61c0a4747f6f2c43e9aaceb810e4775474e29a92477c5039fe619b6235a3ef8aR338